third_party/move/move-vm/runtime/src/execution_tracing/trace.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines the trace data structure which is sufficient to replay Move program execution without
+//! requiring any data accesses (only access to code loader is needed).
+
+use crate::LoadedFunction;
+use bitvec::vec::BitVec;
+use move_core_types::function::ClosureMask;
+use std::rc::Rc;
+
+/// Records the history of conditional branches (taken or not).
+#[derive(Clone)]
+pub(crate) struct CondBrTrace {
+    /// Bit-vector storing the branch history for conditional branches. The vector consists of 64-
+    /// bit blocks that should usually be the fastest on 64-bit CPUs.
+    bits: BitVec<u64>,
+}
+
+impl CondBrTrace {
+    /// Returns an empty branch history.
+    pub(crate) fn empty() -> Self {
+        let bits = BitVec::new();
+        Self { bits }
+    }
+
+    /// Returns an empty history with pre-allocated capacity (number of bits).
+    pub(crate) fn with_capacity(n: usize) -> Self {
+        let bits = BitVec::with_capacity(n);
+        Self { bits }
+    }
+
+    /// Records outcome of a conditional branch.
+    #[inline(always)]
+    pub(crate) fn push(&mut self, taken: bool) {
+        self.bits.push(taken);
+    }
+}
+
+/// A non-static call record in the trace. Used for entry-points and closures.
+#[derive(Clone)]
+pub(crate) enum DynamicCall {
+    Entrypoint(LoadedFunction),
+    Closure(LoadedFunction, ClosureMask),
+}
+
+/// Trace of execution of a program that records information sufficient to replay executed
+/// instructions:
+///   1. Number of successfully executed instructions (ticks).
+///   2. Outcomes of every executed conditional branch.
+///   3. A vector of functions called via closures.
+#[derive(Clone)]
+pub struct Trace {
+    /// Number of successfully executed instructions.
+    ticks: u64,
+    /// Log of all branches taken and not taken.
+    branches: CondBrTrace,
+    /// Log of all functions called via closures. Note that the static calls are not logged to keep
+    /// the log smaller (while giving up the ability to resolve calls without data context when
+    /// replaying the trace).
+    calls: Vec<DynamicCall>,
+}
+
+impl Default for Trace {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl Trace {
+    /// Returns an empty trace.
+    pub fn empty() -> Self {
+        Self {
+            ticks: 0,
+            branches: CondBrTrace::empty(),
+            calls: vec![],
+        }
+    }
+
+    /// Returns a trace from logged data.
+    pub(crate) fn from_logger(ticks: u64, branches: CondBrTrace, calls: Vec<DynamicCall>) -> Self {
+        Self {
+            ticks,
+            branches,
+            calls,
+        }
+    }
+
+    /// Returns true if the trace was fully replayed: all instructions were executed, all branches
```

Comment:
Does it mean it's replayed? It can just be called on newly created empty Trace, right? Or does replay mean something else in this context? in any case, could also just remove this comment I think.